Sources/Plasma/FeatureLib/pfConsoleCore/pfConsoleParser.cpp

```diff
@@ -0,0 +1,157 @@
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include "pfConsoleParser.h"
+
+#include "pfConsoleCmd.h"
+
+static const char kTokenSeparators[] = " =\r\n\t,";
+static const char kTokenGrpSeps[] = " =\r\n._\t,";
+
+std::optional<ST::string> pfConsoleTokenizer::NextNamePart()
+{
+    const char* begin = fPos;
+
+    while (begin != fEnd && isspace(static_cast<unsigned char>(*begin)))
+        ++begin;
+
+    for (fPos = begin; fPos != fEnd; ++fPos) {
+        for (const char *sep = kTokenGrpSeps; *sep; ++sep) {
+            if (*fPos == *sep) {
+                const char* end = fPos;
+                while (fPos != fEnd && (*fPos == *sep)) {
+                    // skip duplicate delimiters
+                    ++fPos;
+                }
+                return ST::string::from_utf8(begin, end - begin);
+            }
+        }
+    }
+
+    if (begin == fPos) {
+        fErrorMsg.clear();
+        return {};
+    }
+
+    return begin;
+}
+
+std::optional<ST::string> pfConsoleTokenizer::NextArgument()
+{
+    const char* begin = fPos;
+
+    while (begin != fEnd && isspace(static_cast<unsigned char>(*begin)))
+        ++begin;
+
+    for (fPos = begin; fPos != fEnd; ++fPos) {
+        if (*begin == '"' || *begin == '\'') {
+            // Handle strings as a single token
+            ++begin;
+            const char* end = begin;
+            while (*end != *fPos) {
+                ++end;
+                if (end == fEnd) {
+                    fErrorMsg = ST_LITERAL("unterminated quoted parameter");
+                    return {};
+                }
+            }
+            fPos = end + 1;
+            return ST::string::from_utf8(begin, end - begin);
+        }
+        for (const char *sep = kTokenSeparators; *sep; ++sep) {
+            if (*fPos == *sep) {
+                const char* end = fPos;
+                while (fPos != fEnd && (*fPos == *sep)) {
+                    // skip duplicate delimiters
+                    ++fPos;
+                }
+                return ST::string::from_utf8(begin, end - begin);
+            }
+        }
+    }
+
+    if (begin == fPos) {
+        fErrorMsg.clear();
+        return {};
+    }
+
+    return begin;
+}
+
+std::pair<pfConsoleCmdGroup*, std::optional<ST::string>> pfConsoleParser::ParseGroupAndName()
```

Comment:
Slight preference for this to be an `std::tuple` because of `std::pair` implying the usage of a map. But that's just a preference because a lot of implementations just have `pair` use `tuple`.

Reply:
Ah hm, is there a convention that `std::pair` is specifically for map-related things? I assumed it was just a specialized alternative to a 2-element `std::tuple`, but I don't know the stdlib too well...